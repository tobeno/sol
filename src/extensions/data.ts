import { ClassWithText } from '../interfaces/base';
import { Constructor } from '../interfaces/util';
import { awaitPromiseSync } from '../utils/async';

export function WithData<T extends Constructor<ClassWithText>>(base: T) {
    return class Wrapped extends base {
        get data(): any {
            if (super.data) {
                return super.data;
            }

            return JSON.parse(this.text);
        }

        set data(value: any) {
            if (super.data) {
                super.data = value;

                return;
            }

            this.text = JSON.stringify(value, null, 2);
        }

        updateData(updater: (value: any) => any | Promise<any>) {
            const updated = awaitPromiseSync(updater(this.data));
            if (typeof updated === 'undefined') {
                throw new Error('Return value missing from updater');
            }

            this.data = updated;
        }

        showData(formatter?: (value: any) => any | Promise<any>) {
            let value = this.data;
            if (formatter) {
                value = awaitPromiseSync(formatter(value));
            }

            console.log(value);
        }
    };
}
