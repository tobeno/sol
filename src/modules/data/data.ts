import jsonata, { Expression } from 'jsonata';
import { transform, wrapString } from './transformer';
import { Wrapper } from './wrapper';
import { Text } from './text';
import { DataSource } from './data-source';
import { inspect } from 'util';
import { DataTransformation } from './data-transformation';
import type {
  AnyItemType,
  AnyKeyType,
  AnyPartial,
} from '../../interfaces/util';
import {
  camelcaseObject,
  constantcaseObject,
  snakecaseObject,
} from '../utils/object';
import { DataType } from './data-type';

/**
 * Generic wrapper for runtime objects
 */
export class Data<
  ValueType = any,
  ItemType = AnyItemType<ValueType>,
  KeyType = AnyKeyType<ValueType>,
> extends Wrapper<ValueType> {
  constructor(
    value: ValueType,
    public source: DataSource | null = null,
    public sourceTransformation: DataTransformation | null = null,
  ) {
    super(value);
  }

  get rootSource(): DataSource | null {
    if (!this.source) {
      return null;
    }

    return this.source.rootSource || this.source;
  }

  get clone(): Data<ValueType> {
    return new Data<ValueType>(JSON.parse(JSON.stringify(this.value)), this);
  }

  get flattened(): Data<
    ValueType extends Array<any> ? FlatArray<ValueType, 1> : ValueType
  > {
    if (Array.isArray(this.value)) {
      return new Data(this.value.flat() as any, this) as any;
    }

    return this as any;
  }

  get first(): Data<ItemType> | null {
    if (Array.isArray(this.value)) {
      return this.value.length ? new Data<ItemType>(this.value[0], this) : null;
    }

    const values = Object.values(this.value);

    return values.length ? new Data<ItemType>(values[0], this) : null;
  }

  get last(): Data<ItemType> | null {
    if (Array.isArray(this.value)) {
      return this.value.length
        ? new Data<ItemType>(this.value[this.value.length - 1], this)
        : null;
    }

    const values = Object.values(this.value);

    return values.length
      ? new Data<ItemType>(values[values.length - 1], this)
      : null;
  }

  get sum(): Data<number> {
    const values = this.values;

    return values.reduce(
      (result, value) => result + (parseFloat(value as any) || 0),
      0,
    );
  }

  get avg(): Data<number> {
    const values = this.values;

    return new Data(values.sum.value / values.length);
  }

  get unique(): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return new Data(this.value.unique, this) as any;
    }

    const knownValues: any[] = [];

    return this.filter((value: any) => {
      const result = !knownValues.includes(value);
      knownValues.push(value);
      return result;
    });
  }

  get sorted(): Data<ValueType extends Array<any> ? ValueType : any> {
    return this.sort();
  }

  get reversed(): Data<ValueType extends Array<any> ? ValueType : any> {
    if (Array.isArray(this.value)) {
      return new Data(this.value.reverse(), this) as any;
    }

    return new Data(
      Object.fromEntries(Object.entries(this.value).reverse()),
      this,
    ) as any;
  }

  get filtered(): Data<AnyPartial<ValueType>> {
    return this.filter((value: any) => !!value);
  }

  get camelcased(): Data<any> {
    return new Data(camelcaseObject(this.value), this) as any;
  }

  get snakecased(): Data<any> {
    return new Data(snakecaseObject(this.value), this) as any;
  }

  get constantcased(): Data<any> {
    return new Data(constantcaseObject(this.value), this) as any;
  }

  get keys(): Data<KeyType[]> {
    if (Array.isArray(this.value)) {
      return new Data(
        (this.value as any).map((_: any, index: any) => index),
      ) as any;
    }

    return new Data(Object.keys(this.value)) as any;
  }

  get values(): Data<ItemType[]> {
    if (Array.isArray(this.value)) {
      return new Data(this.value) as any;
    }

    return new Data(Object.values(this.value));
  }

  get length(): number {
    return (this.value as any).length;
  }

  get entries(): Data<[KeyType, ItemType][]> {
    if (Array.isArray(this.value)) {
      return new Data(this.value.entries()) as any;
    }

    return new Data(Object.entries(this.value)) as any;
  }

  setSource(source: DataSource | null): Data<ValueType> {
    this.source = source;

    return this as any;
  }

  group<KeyType extends string | number | symbol, GroupType = ItemType[]>(
    keyFn: (item: ItemType) => KeyType,
    reduceFn: (group: GroupType, item: ItemType) => GroupType = (
      group,
      item,
    ) => [...group, item],
    groupFn: (key: KeyType) => GroupType,
  ): Data<Record<KeyType, ItemType[]>> {
    const values = this.values;

    return values.reduce((result, item) => {
      const key = keyFn(item);
      let group = result[key];
      if (!group) {
        group = [];
        result[key] = group;
      }

      group.push(item);

      return result;
    }, {} as Record<KeyType, ItemType[]>);
  }

  sort(
    compareFn?: (a: ItemType, b: ItemType) => number,
  ): Data<ValueType extends Array<any> ? ValueType : any> {
    if (Array.isArray(this.value)) {
      return new Data([...this.value].sort(compareFn), this) as any;
    }

    const entries = this.entries.value;
    entries.sort(([aIndex, a], [bIndex, b]) => {
      if (compareFn) {
        return compareFn(a, b);
      }

      if (a < b) {
        return -1;
      }

      if (a > b) {
        return 1;
      }

      return 0;
    });

    return new Data(Object.fromEntries(entries), this) as any;
  }

  filter(
    cb: (value: ItemType, index: KeyType) => boolean,
  ): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return new Data(this.value.filter(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        newValue[index] = value;
      }
    });

    return new Data(newValue, this) as any;
  }

  find(
    cb: (value: ItemType, index: KeyType) => boolean,
  ): Data<ItemType> | undefined {
    if (Array.isArray(this.value)) {
      const result = this.value.find(cb as any) as any;

      return result ? (new Data(result, this) as any) : undefined;
    }

    let result: any = null;
    for (const index of this.keys.value) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = value;

        break;
      }
    }

    return result ? (new Data(result, this) as any) : undefined;
  }

  findIndex(cb: (value: ItemType, index: KeyType) => boolean): KeyType {
    if (Array.isArray(this.value)) {
      return this.value.findIndex(cb as any) as any;
    }

    let result: any = null;
    for (const index of this.keys.value) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = index;

        break;
      }
    }

    return result;
  }

  reduce<ResultType>(
    cb: (result: ResultType, item: ItemType, index: KeyType) => ResultType,
    initial: ResultType,
  ): Data<ResultType> {
    if (Array.isArray(this.value)) {
      return new Data<ResultType>(
        this.value.reduce(
          (result, value, index) => cb(result, value, index as any),
          initial,
        ),
        this,
      );
    }

    return new Data(
      this.entries.value.reduce(
        (result, [key, value]) => cb(result, value, key),
        initial,
      ),
      this,
    );
  }

  some(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.some(cb as any);
    }

    let result = false;
    for (const index of this.keys.value) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = true;

        break;
      }
    }

    return result;
  }

  every(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.every(cb as any);
    }

    let result = true;
    for (const index of this.keys.value) {
      const value = (this.value as any)[index];
      if (!cb(value, index)) {
        result = false;

        break;
      }
    }

    return result;
  }

  map<MappedItemType = any>(
    cb: (value: ItemType, index: KeyType) => MappedItemType,
  ): Data<
    ValueType extends Array<any>
      ? MappedItemType[]
      : ValueType extends Record<string, any>
      ? { [key: string]: MappedItemType }
      : any
  > {
    if (Array.isArray(this.value)) {
      return new Data((this.value as any).map(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      newValue[index] = cb(value, index);
    });

    return new Data(newValue, this) as any;
  }

  join(separator: string): Text {
    if (Array.isArray(this.value)) {
      return wrapString(this.value.join(separator), null, this);
    }

    return wrapString(Object.values(this.value).join(separator), null, this);
  }

  forEach(cb: (value: ItemType, index: KeyType) => any) {
    if (Array.isArray(this.value)) {
      return this.value.forEach(cb as any);
    }

    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      cb(value, index);
    });
  }

  extract<ExtractedValueType = any>(
    exp: string | Expression,
  ): Data<ExtractedValueType> {
    if (typeof exp === 'string') {
      exp = jsonata(exp);
    }

    return new Data(exp.evaluate(this.value), this) as any;
  }

  transformTo<TargetType = any>(targetType: DataType | string): TargetType {
    if (typeof targetType === 'string') {
      targetType = DataType.fromString(targetType);
    }

    return transform(this, new DataTransformation(DataType.Data, targetType));
  }

  [inspect.custom]() {
    return `Data ${inspect(this.value)}`;
  }
}
