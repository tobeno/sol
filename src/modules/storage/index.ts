export * from './utils/save.utils';
