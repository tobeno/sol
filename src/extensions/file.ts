import { Constructor } from '../interfaces/util';
import { File } from '../storage/file';

export function WithFile<T extends Constructor>(base: T) {
  return class Wrapped extends base {
    file(pathOrFile: string | File): File {
      const f = pathOrFile instanceof File ? pathOrFile : new File(pathOrFile);
      f.text = String(this);
      return f;
    }
  };
}
