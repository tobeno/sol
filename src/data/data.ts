import { Expression } from 'jsonata';
import * as jsonata from 'jsonata';
import {
  wrapString,
  dataToJson,
  dataToYaml,
  dataToCsv,
  wrapObject,
} from './mapper';
import { Wrapper } from './wrapper';
import { Text } from './text';
import { DataSource } from './data-source';
import { inspect } from 'util';
import { DataTransformation } from './data-transformation';
import { clipboard } from '../os/clipboard';
import { AnyItemType, AnyKeyType, AnyPartial } from '../interfaces/util';
import { edit } from '../integrations/editor';
import { File } from '../storage/file';
import { save, saveAs } from '../storage/save';
import { camelcaseObject, snakecaseObject } from '../utils/object';

/**
 * Generic wrapper for runtime objects
 */
export class Data<
  ValueType = any,
  ItemType = AnyItemType<ValueType>,
  KeyType = AnyKeyType<ValueType>
> extends Wrapper<ValueType> {
  constructor(
    value: ValueType,
    public source: DataSource | null = null,
    public sourceTransformation: DataTransformation | null = null,
  ) {
    super(value);
  }

  get rootSource(): DataSource | null {
    if (!this.source) {
      return null;
    }

    return this.source.rootSource || this.source;
  }

  get text(): Text<ValueType> {
    let value = this.value as any;
    if (!(value instanceof Text) && typeof value !== 'string') {
      value = this.json;
    }

    return wrapString(value, null, this);
  }

  get json(): Text<ValueType> {
    return dataToJson<ValueType>(this);
  }

  get yaml(): Text<ValueType> {
    return dataToYaml<ValueType>(this);
  }

  get csv(): Text<ValueType> {
    return dataToCsv<ValueType>(this);
  }

  get clone(): Data<ValueType> {
    return wrapObject<ValueType>(JSON.parse(JSON.stringify(this.value)), this);
  }

  get flattened(): Data<
    ValueType extends Array<any> ? FlatArray<ValueType, 1> : ValueType
  > {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.flat() as any, this) as any;
    }

    return this as any;
  }

  get unique(): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return wrapObject([...new Set(this.value)] as any, this) as any;
    }

    const knownValues: any[] = [];

    return this.filter((value: any) => {
      const result = !knownValues.includes(value);
      knownValues.push(value);
      return result;
    });
  }

  get sorted(): Data<ValueType extends Array<any> ? ValueType : any> {
    return this.sort();
  }

  get filtered(): Data<AnyPartial<ValueType>> {
    return this.filter((value: any) => !!value);
  }

  get camelcased(): Data<any> {
    return wrapObject(camelcaseObject(this.value), this) as any;
  }

  get snakecased(): Data<any> {
    return wrapObject(snakecaseObject(this.value), this) as any;
  }

  get keys(): KeyType[] {
    if (Array.isArray(this.value)) {
      return (this.value as any).map((_: any, index: any) => index) as any;
    }

    return Object.keys(this.value) as any;
  }

  get values(): ItemType[] {
    if (Array.isArray(this.value)) {
      return this.value as any;
    }

    return Object.values(this.value);
  }

  get entries(): [KeyType, ItemType][] {
    if (Array.isArray(this.value)) {
      return this.value.entries() as any;
    }

    return Object.entries(this.value) as any;
  }

  withSource(source: DataSource | null): Data<ValueType> {
    this.source = source;

    return this as any;
  }

  sort(
    compareFn?: (a: ItemType, b: ItemType) => number,
  ): Data<ValueType extends Array<any> ? ValueType : any> {
    if (Array.isArray(this.value)) {
      return wrapObject([...this.value].sort(compareFn), this) as any;
    }

    const entries = this.entries;
    entries.sort(([aIndex, a], [bIndex, b]) => {
      if (compareFn) {
        return compareFn(a, b);
      }

      if (a < b) {
        return -1;
      }

      if (a > b) {
        return 1;
      }

      return 0;
    });

    return wrapObject(Object.fromEntries(entries), this) as any;
  }

  filter(
    cb: (value: ItemType, index: KeyType) => boolean,
  ): Data<AnyPartial<ValueType>> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.filter(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        newValue[index] = value;
      }
    });

    return wrapObject(newValue, this) as any;
  }

  find(cb: (value: ItemType, index: KeyType) => boolean): Data<ItemType> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.find(cb as any) as any, this) as any;
    }

    let result: any = null;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = value;

        break;
      }
    }

    return result ? (wrapObject(result, this) as any) : null;
  }

  findIndex(cb: (value: ItemType, index: KeyType) => boolean): Data<KeyType> {
    if (Array.isArray(this.value)) {
      return wrapObject(this.value.findIndex(cb as any) as any, this) as any;
    }

    let result: any = null;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = index;

        break;
      }
    }

    return result ? (wrapObject(result, this) as any) : null;
  }

  some(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.some(cb as any);
    }

    let result = false;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (cb(value, index)) {
        result = true;

        break;
      }
    }

    return result;
  }

  every(cb: (value: ItemType, index: KeyType) => boolean): boolean {
    if (Array.isArray(this.value)) {
      return this.value.every(cb as any);
    }

    let result = true;
    for (const index of this.keys) {
      const value = (this.value as any)[index];
      if (!cb(value, index)) {
        result = false;

        break;
      }
    }

    return result;
  }

  map<MappedItemType = any>(
    cb: (value: ItemType, index: KeyType) => MappedItemType,
  ): Data<
    ValueType extends Array<any>
      ? MappedItemType[]
      : ValueType extends Record<string, any>
      ? { [key: string]: MappedItemType }
      : any
  > {
    if (Array.isArray(this.value)) {
      return wrapObject((this.value as any).map(cb as any) as any, this) as any;
    }

    const newValue: any = {};
    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      newValue[index] = cb(value, index);
    });

    return wrapObject(newValue, this) as any;
  }

  join(separator: string): Text {
    if (Array.isArray(this.value)) {
      return wrapString(this.value.join(separator), null, this);
    }

    return wrapString(Object.values(this.value).join(separator), null, this);
  }

  forEach(cb: (value: ItemType, index: KeyType) => any) {
    if (Array.isArray(this.value)) {
      return this.value.forEach(cb as any);
    }

    this.keys.forEach((index: any) => {
      const value = (this.value as any)[index];
      cb(value, index);
    });
  }

  transform<TransformedValueType = any>(
    exp: string | Expression,
  ): Data<TransformedValueType> {
    if (typeof exp === 'string') {
      exp = jsonata(exp);
    }

    return wrapObject(exp.evaluate(this.value), this) as any;
  }

  copy() {
    clipboard.text = String(this);
  }

  edit(): File {
    return edit(this);
  }

  print() {
    console.log(String(this));
  }

  saveAs(path: string): File {
    return saveAs(this, path);
  }

  save(): File {
    return save(this);
  }

  [inspect.custom]() {
    return `Data ${inspect(this.value)}`;
  }
}
