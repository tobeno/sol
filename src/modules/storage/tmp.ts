import { fileSync as tmpFileSync } from 'tmp';
import { File } from './file';

/**
 * Returns a temporary file.
 */
export function tmp(ext?: string): File {
  return File.create(
    tmpFileSync({
      ...(ext ? { postfix: `.${ext}` } : {}),
    }).name,
  );
}
