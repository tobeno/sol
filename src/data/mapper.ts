import { AnyMapper } from './mappers/any-mapper';
import { DataTransformation } from './data-transformation';
import { WrappingMapper } from './mappers/wrapping-mapper';
import { Data } from './data';
import { DataType } from './data-type';
import { DataFormat } from './data-format';
import { CsvMapper } from './mappers/csv-mappers';
import { JsonMapper } from './mappers/json-mappers';
import { YamlMapper } from './mappers/yaml-mappers';
import { HtmlMapper } from './mappers/html-mappers';
import { XmlMapper } from './mappers/xml-mappers';
import { TextCommaSeparatedMapper } from './mappers/text-comma-separated-mapper';
import { TextNewlineSeparatedMapper } from './mappers/text-newline-separated-mapper';
import { TextSemicolonSeparatedMapper } from './mappers/text-semicolon-separated-mapper';
import { ToStringMapper } from './mappers/tostring-mapper';
import { Html } from './html';
import { Xml } from './xml';
import { Text } from './text';
import { TextMapper } from './mappers/text-mapper';
import { DataSourceMapper } from './mappers/data-source-mapper';
import { DataSource } from './data-source';
import { DataTransformationMapper } from './mappers/data-transformation-mapper';
import { Ast } from './ast';
import { AstMapper } from './mappers/ast-mapper';

const anyMapper = new AnyMapper([
  new AstMapper(),
  new CsvMapper(),
  new JsonMapper(),
  new YamlMapper(),
  new HtmlMapper(),
  new XmlMapper(),
  new TextMapper(),
  new TextCommaSeparatedMapper(),
  new TextNewlineSeparatedMapper(),
  new TextSemicolonSeparatedMapper(),
  new ToStringMapper(),
]);
export const mapper = new DataTransformationMapper(
  new DataSourceMapper(new WrappingMapper(Data, anyMapper)),
);

export function map(
  input: any,
  transformation: DataTransformation | string,
): any {
  if (typeof transformation === 'string') {
    transformation = DataTransformation.fromString(transformation);
  }

  return mapper.map(input, transformation);
}

export function jsonToData(
  value: string | String,
  source: DataSource | null = null,
): Data {
  source = source || (value as any).source || null;

  return map(
    unwrapString(value),
    new DataTransformation(
      DataType.String.withFormat(DataFormat.Json),
      DataType.Data,
    ),
  ).withSource(source);
}

export function dataToJson(
  value: Data | any,
  source: DataSource | null = null,
): Text {
  if (!(value instanceof Data)) {
    value = new Data(value);
  }

  source = source || value.source;

  return wrapString(
    map(
      value,
      new DataTransformation(
        DataType.Data,
        DataType.String.withFormat(DataFormat.Json),
      ),
    ),
    DataFormat.Json,
  ).withSource(source);
}

export function yamlToData(
  value: string | String | Text,
  source: DataSource | null = null,
): Data {
  source = source || (value as any).source || null;

  return map(
    unwrapString(value),
    new DataTransformation(
      DataType.String.withFormat(DataFormat.Yaml),
      DataType.Data,
    ),
  ).withSource(source);
}

export function dataToYaml(
  value: Data | any,
  source: DataSource | null = null,
): Text {
  if (!(value instanceof Data)) {
    value = new Data(value);
  }

  source = source || value.source;

  return wrapString(
    map(
      value,
      new DataTransformation(
        DataType.Data,
        DataType.String.withFormat(DataFormat.Yaml),
      ),
    ),
    DataFormat.Yaml,
  ).withSource(source);
}

export function csvToData(
  value: string | String,
  source: DataSource | null = null,
): Data {
  source = source || (value as any).source || null;

  return map(
    unwrapString(value),
    new DataTransformation(
      DataType.String.withFormat(DataFormat.Csv),
      DataType.Data,
    ),
  );
}

export function dataToCsv(
  value: Data | any,
  source: DataSource | null = null,
): Text {
  if (!(value instanceof Data)) {
    value = new Data(value);
  }

  source = source || value.source;

  return wrapString(
    map(
      value,
      new DataTransformation(
        DataType.Data,
        DataType.String.withFormat(DataFormat.Csv),
      ),
    ),
    DataFormat.Csv,
  ).withSource(source);
}

export function codeToAst(
  value: string | String,
  source: DataSource | null = null,
): Ast {
  source = source || (value as any).source || null;

  return map(
    unwrapString(value),
    new DataTransformation(DataType.String, DataType.Ast),
  );
}

export function astToCode(
  value: Ast | any,
  source: DataSource | null = null,
): Text {
  if (!(value instanceof Ast)) {
    value = new Ast(value);
  }

  source = source || value.source;

  return wrapString(
    map(value, new DataTransformation(DataType.Ast, DataType.String)),
    DataFormat.Csv,
  ).withSource(source);
}

export function wrapString(
  value: string | String,
  format: string | null = null,
  source: DataSource | null = null,
): Text {
  if (value instanceof Text) {
    let text = value;

    if (format) {
      text = text.withFormat(format);
    }

    if (source) {
      text = text.withSource(source);
    }

    return text;
  }

  return new Text(value, format, source);
}

export function unwrapString(value: string | String | Text): string {
  if (value instanceof Text) {
    value = value.value;
  }

  return value.toString();
}

export function wrapObject(value: any, source: DataSource | null = null): Data {
  return new Data(value, source);
}

export function unwrapObject(value: any | Data): any {
  if (value instanceof Data) {
    value = value.value;
  }

  return value;
}

export function wrapHtml(
  value: string | String,
  source: DataSource | null = null,
): Html {
  return map(
    unwrapString(value),
    new DataTransformation(
      DataType.String.withFormat(DataFormat.Html),
      DataType.Html,
    ),
  ).withSource(source);
}

export function unwrapHtml(value: Html): Text {
  return wrapString(value.value, DataFormat.Html, value.source);
}

export function wrapXml(
  value: string | String,
  source: DataSource | null = null,
): Xml {
  return map(
    unwrapString(value),
    new DataTransformation(
      DataType.String.withFormat(DataFormat.Xml),
      DataType.Xml,
    ),
  ).withSource(source);
}

export function unwrapXml(value: Xml): Text {
  return wrapString(value.value, DataFormat.Xml, value.source);
}
