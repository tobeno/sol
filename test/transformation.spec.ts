import * as tmp from 'tmp';
import { file } from '../src/storage/file';

import { readFileSync } from 'fs';
import { wrapObject } from '../src/data/mapper';

describe('transformation', () => {
  it('should transform a JSON file', async () => {
    const data = file(`${__dirname}/assets/products.json`).json;
    const obj = data.value;

    expect(obj).toHaveProperty('products');

    obj.variants = Array.from(
      wrapObject(obj.products).transform('variants').value,
    );

    const tmpFile = tmp.fileSync();
    data.json.saveAs(tmpFile.name);
    expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
    tmpFile.removeCallback();

    const tmpFile2 = tmp.fileSync();
    data.transform('variants').csv.saveAs(tmpFile2.name);
    expect(readFileSync(tmpFile2.name, 'utf-8')).toMatchSnapshot();
    tmpFile2.removeCallback();
  });

  it('should transform a YAML file', async () => {
    const data = file(`${__dirname}/assets/products.yaml`).yaml;
    const obj = data.value;

    expect(obj).toHaveProperty('products');

    const tmpFile = tmp.fileSync();
    data.yaml.saveAs(tmpFile.name);
    expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
    tmpFile.removeCallback();
  });

  it('should transform a CSV file', async () => {
    let data = file(`${__dirname}/assets/variants.csv`).csv;

    data = data.sort((a, b) => a.quantity - b.quantity);

    const tmpFile = tmp.fileSync();
    data.csv.saveAs(tmpFile.name);
    expect(readFileSync(tmpFile.name, 'utf-8')).toMatchSnapshot();
    tmpFile.removeCallback();
  });
});
