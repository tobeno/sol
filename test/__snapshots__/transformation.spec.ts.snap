// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformation should transform a CSV file 1`] = `
"id,size,quantity
1-3,44,1
2-3,38,1
2-1,36,2
2-2,37,2
1-1,42,3
1-2,43,20"
`;

exports[`transformation should transform a JSON file 1`] = `
"{
  \\"products\\": [
    {
      \\"id\\": \\"1\\",
      \\"name\\": \\"Leather shoe\\",
      \\"variants\\": [
        {
          \\"id\\": \\"1-1\\",
          \\"size\\": \\"42\\",
          \\"quantity\\": 3
        },
        {
          \\"id\\": \\"1-2\\",
          \\"size\\": \\"43\\",
          \\"quantity\\": 20
        },
        {
          \\"id\\": \\"1-3\\",
          \\"size\\": \\"44\\",
          \\"quantity\\": 1
        }
      ]
    },
    {
      \\"id\\": \\"2\\",
      \\"name\\": \\"Stiletto shoe\\",
      \\"variants\\": [
        {
          \\"id\\": \\"2-1\\",
          \\"size\\": \\"36\\",
          \\"quantity\\": 2
        },
        {
          \\"id\\": \\"2-2\\",
          \\"size\\": \\"37\\",
          \\"quantity\\": 2
        },
        {
          \\"id\\": \\"2-3\\",
          \\"size\\": \\"38\\",
          \\"quantity\\": 1
        }
      ]
    }
  ],
  \\"variants\\": [
    {
      \\"id\\": \\"1-1\\",
      \\"size\\": \\"42\\",
      \\"quantity\\": 3
    },
    {
      \\"id\\": \\"1-2\\",
      \\"size\\": \\"43\\",
      \\"quantity\\": 20
    },
    {
      \\"id\\": \\"1-3\\",
      \\"size\\": \\"44\\",
      \\"quantity\\": 1
    },
    {
      \\"id\\": \\"2-1\\",
      \\"size\\": \\"36\\",
      \\"quantity\\": 2
    },
    {
      \\"id\\": \\"2-2\\",
      \\"size\\": \\"37\\",
      \\"quantity\\": 2
    },
    {
      \\"id\\": \\"2-3\\",
      \\"size\\": \\"38\\",
      \\"quantity\\": 1
    }
  ]
}"
`;

exports[`transformation should transform a JSON file 2`] = `
"id,size,quantity
1-1,42,3
1-2,43,20
1-3,44,1
2-1,36,2
2-2,37,2
2-3,38,1"
`;

exports[`transformation should transform a YAML file 1`] = `
"products:
  - id: \\"1\\"
    name: Leather shoe
    variants:
      - id: 1-1
        size: \\"42\\"
        quantity: 3
      - id: 1-2
        size: \\"43\\"
        quantity: 20
      - id: 1-3
        size: \\"44\\"
        quantity: 1
  - id: \\"2\\"
    name: Stiletto shoe
    variants:
      - id: 2-1
        size: \\"36\\"
        quantity: 2
      - id: 2-2
        size: \\"37\\"
        quantity: 2
      - id: 2-3
        size: \\"38\\"
        quantity: 1
"
`;
